### PR TITLE
Raise workspace events outside of serialization lock

### PR DESCRIPTION
**Pull request by tmat**
Fixes https://github.com/dotnet/roslyn/issues/42229

### CODE REVIEW

File: src/Workspaces/Core/Portable/Workspace/Workspace.cs [hunk 1]
```diff
@@ -352,38 +360,44 @@ protected virtual void Dispose(bool finalize)
...
+                newSolution = this.SetCurrentSolution(this.CreateSolution(reloadedSolutionInfo));
 
-                reloadedSolutionInfo.Projects.Do(pi => OnProjectAdded_NoLock(pi, silent: true));
+                reloadedSolutionInfo.Projects.Do(pi => OnProjectAdded_NoLock(pi));
 
                 newSolution = this.AdjustReloadedSolution(oldSolution, this.CurrentSolution);
                 newSolution = this.SetCurrentSolution(newSolution);
 
-                this.RaiseWorkspaceChangedEventAsync(WorkspaceChangeKind.SolutionReloaded, oldSolution, newSolution);
             }
```
Comment (genlu) [inline src/Workspaces/Core/Portable/Workspace/Workspace.cs]: extra newline above (and a few other places in this file)

File: src/Workspaces/Core/Portable/Workspace/Workspace.cs [hunk 2]
```diff
@@ -735,71 +758,78 @@ protected virtual void CheckDocumentCanBeRemoved(DocumentId documentId)
...
+                oldSolution = this.CurrentSolution;
 
-                var newSolution = oldSolution.WithAnalyzerConfigDocumentTextLoader(documentId, loader, PreservationMode.PreserveValue);
+                newSolution = oldSolution.WithAnalyzerConfigDocumentTextLoader(documentId, loader, PreservationMode.PreserveValue);
                 newSolution = this.SetCurrentSolution(newSolution);
-
-                this.RaiseWorkspaceChangedEventAsync(WorkspaceChangeKind.AnalyzerConfigDocumentChanged, oldSolution, newSolution, documentId: documentId);
             }
+
+            this.RaiseWorkspaceChangedEventAsync(WorkspaceChangeKind.AnalyzerConfigDocumentChanged, oldSolution, newSolution, documentId: documentId);
```
Comment (mavasani) [inline src/Workspaces/Core/Portable/Workspace/Workspace.cs]: Can you add a comment for each method where we moved this out so it does not regress? Any change we can write some unit tests?

**Comment by genlu** [comment]
so now the events can be raised in different order than actual solution change, could that break the listeners? Or this was never a guarantee?

**Comment by tmat** [comment]
So, this actually is not enough. There is still a deadlock - `_serializationLock` is also taken in OnDocumentClosed. :(

```
> 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Workspace.UpdateCurrentSolutionOnOptionsChanged() Line 237	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Options.GlobalOptionService.UpdateRegisteredWorkspacesAndRaiseEvents(System.Collections.Generic.List<Microsoft.CodeAnalysis.Options.OptionChangedEventArgs> changedOptions) Line 364	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Options.GlobalOptionService.SetOptions(Microsoft.CodeAnalysis.Options.OptionSet optionSet) Line 325	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Options.OptionServiceFactory.OptionService.SetOptions(Microsoft.CodeAnalysis.Options.OptionSet optionSet) Line 128	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Workspace.SetOptions(Microsoft.CodeAnalysis.Options.OptionSet options) Line 231	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Workspace.TryApplyChanges(Microsoft.CodeAnalysis.Solution newSolution, Microsoft.CodeAnalysis.Shared.Utilities.IProgressTracker progressTracker) Line 1255	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Workspace.TryApplyChanges(Microsoft.CodeAnalysis.Solution newSolution) Line 1194	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.DiagnosticProvider.Enable(Microsoft.CodeAnalysis.Workspace workspace, Microsoft.CodeAnalysis.DiagnosticProvider.Options options) Line 21	C#
 	Microsoft.CodeAnalysis.Features.dll!Microsoft.CodeAnalysis.SolutionCrawler.MiscSolutionCrawlerWorkspaceEventListener.StartListening(Microsoft.CodeAnalysis.Workspace workspace, object serviceOpt) Line 23	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Host.DefaultWorkspaceEventListenerServiceFactory.Service.EnsureListeners() Line 74	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Workspace.GetEventHandlers<Microsoft.CodeAnalysis.WorkspaceChangeEventArgs>(string eventName) Line 216	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Workspace.RaiseWorkspaceChangedEventAsync(Microsoft.CodeAnalysis.WorkspaceChangeKind kind, Microsoft.CodeAnalysis.Solution oldSolution, Microsoft.CodeAnalysis.Solution newSolution, Microsoft.CodeAnalysis.ProjectId projectId, Microsoft.CodeAnalysis.DocumentId documentId) Line 58	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Workspace.OnProjectAdded(Microsoft.CodeAnalysis.ProjectInfo projectInfo) Line 443	C#
 	Microsoft.VisualStudio.LanguageServices.dll!Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.MiscellaneousFilesWorkspace.AttachToDocument(string moniker, Microsoft.VisualStudio.Text.ITextBuffer textBuffer) Line 262	C#
 	Microsoft.VisualStudio.LanguageServices.dll!Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.MiscellaneousFilesWorkspace.Registration_WorkspaceChanged(object sender, System.EventArgs e) Line 203	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.WorkspaceRegistration.RaiseEvents() Line 53	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.WorkspaceRegistration.SetWorkspaceAndRaiseEvents(Microsoft.CodeAnalysis.Workspace workspace) Line 26	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Workspace.UnregisterText(Microsoft.CodeAnalysis.Text.SourceTextContainer textContainer) Line 65	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Workspace.ClearOpenDocument(Microsoft.CodeAnalysis.DocumentId documentId) Line 100	C#
>	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Workspace.OnDocumentClosed(Microsoft.CodeAnalysis.DocumentId documentId, Microsoft.CodeAnalysis.TextLoader reloader, bool updateActiveContext) Line 572	C#
 	Microsoft.VisualStudio.LanguageServices.dll!Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.VisualStudioWorkspaceImpl.OpenFileTracker.TryClosingDocumentsForMoniker.AnonymousMethod__0(Microsoft.CodeAnalysis.Workspace w) Line 317	C#

```

**Comment by tmat** [comment]
@CyrusNajmabadi What's the benefit of that? I want to remove the lock completely at some point.

**Comment by tmat** [comment]
@genlu 
> so now the events can be raised in different order than actual solution change, could that break the listeners? Or this was never a guarantee?

I don't think there is any guarantee of ordering. If two threads issued a changed to the workspace in parallel then even previously the order in which these changes are applied and the events raised is non-deterministic. Each even has `oldSolution` and `newSolution` and each `Solution` snapshot has a `WorkspaceVersion`. The version can be used for ordering.

**Comment by genlu** [comment]
> If two threads issued a changed to the workspace in parallel then even previously the order in which these changes are applied and the events raised is non-deterministic

That's true, but the lock still guarantees that the later event would contain newer "new solution" than previous ones, which is not the case after this change. I guess my point is we need to check components depend on workspace event don't rely on the ordering or use `WorkspaceVersion` properly.

**Comment by tmat** [comment]
@genlu I see. Makes sense. We should order the notifications to match the order of updates.

**Pull request closed without merging**